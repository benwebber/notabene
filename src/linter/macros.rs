macro_rules! invalid_span {
    ($check:ident) => {
        #[derive(Default)]
        pub struct $check;

        impl Check for $check {
            fn rule(&self) -> Rule {
                Rule::$check
            }

            fn visit_invalid_span(&mut self, context: &mut Context, span: &parsed::InvalidSpan) {
                if let parsed::InvalidSpan::$check(s) = span {
                    context.report(self.rule(), Some(*s));
                }
            }
        }
    };
}
