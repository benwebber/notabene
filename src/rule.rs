//! Linter rules.
use serde::{Deserialize, Serialize};

macro_rules! rules {
    ($($rule:ident = ($doc:literal, $code:literal, $message:literal $(,)?)),* $(,)?) => {
        /// A linter rule.
        #[derive(Clone, Copy, Debug, PartialEq, Deserialize, Serialize)]
        #[non_exhaustive]
        pub enum Rule {
            $(#[doc = concat!("`", $code, "`. ")] #[doc = $doc] $rule),*
        }

        impl Rule {
            /// All rules.
            pub const ALL: [Self; [$(stringify!($rule)),*].len()] = [
                $(Self::$rule),*
            ];

            /// Return the error code for this rule.
            pub fn code(&self) -> &str {
                match self {
                    $(Rule::$rule => $code),*
                }
            }

            /// Return the documentation this rule.
            pub fn doc(&self) -> &str {
                match self {
                    $(Rule::$rule => $doc),*
                }
            }

            /// Return the message template for this rule.
            pub fn message(&self) -> &str {
                match self {
                    $(Rule::$rule => $message),*
                }
            }
        }
    };
}

rules! {
    // E000 Document
    MissingTitle = (
        "The title is missing.",
        "E001",
        "Missing title",
    ),
    InvalidTitle = (
        "The title is not plain text.",
        "E002",
        "Invalid title `{}`",
    ),
    DuplicateTitle = (
        "There is a duplicate `h1` in the document.",
        "E003",
        "Duplicate title `{}`",
    ),
    InvalidSectionHeading = (
        "The `h2` is not a valid unreleased or release section heading.",
        "E004",
        "Invalid heading `{}`",
    ),
    // E100 Unreleased
    MissingUnreleased = (
        "The document does not have an unreleased section.",
        "E100",
        "Missing unreleased heading",
    ),
    DuplicateUnreleased = (
        "There is more than one unreleased section heading in the document.",
        "E101",
        "Duplicate unreleased section `{}`",
    ),
    // E200 Release
    InvalidDate = (
        "The date is not in ISO 8601 format.",
        "E200",
        "Invalid date `{}`",
    ),
    InvalidYanked = (
        "The yanked token does not match `[YANKED]`.",
        "E201",
        "Invalid [YANKED] format `{}`",
    ),
    // E300 Changes
    InvalidChangeType = (
        "The change section heading is not a known change type.",
        "E300",
        "Invalid change type `{}`",
    ),
    DuplicateChangeType = (
        "There is more than one change section with the same change type.",
        "E301",
        "Duplicate change type `{}`",
    ),
    // E400 Content
    EmptySection = (
        "A section is unexpectedly empty (e.g. a release with no changes).",
        "E400",
        "Empty section",
    ),
}

#[cfg(test)]
mod tests {
    use super::*;

    use std::collections::HashSet;

    #[test]
    fn test_rule_codes_unique() {
        let mut codes = HashSet::new();
        let mut duplicates: Vec<&str> = vec![];
        for rule in Rule::ALL.iter() {
            if !codes.insert(rule.code()) {
                duplicates.push(rule.code());
            }
        }
        assert_eq!(duplicates, Vec::<&str>::new());
    }
}
