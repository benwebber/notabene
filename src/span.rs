//! Work with spans within the source document.
use std::ops::{Deref, Range};

use serde::{Deserialize, Serialize};

mod iter;
mod locator;
mod ranged;
mod unist;

pub(crate) use iter::SpanIterator;
pub use locator::Locator;
pub(crate) use ranged::Ranged;
pub use unist::{Point, Position};

/// A span within the source document.
///
/// Use [`Locator`] to convert a `Span` to a [`Position`].
#[derive(
    Copy, Clone, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize,
)]
pub struct Span {
    /// The start offset of the span.
    pub start: usize,
    /// The end offset of the span.
    pub end: usize,
}

#[derive(Clone, Copy, Debug, Default, Eq, Hash, PartialEq, Serialize)]
pub(crate) struct Spanned<T> {
    pub span: Span,
    pub value: T,
}

impl Ranged<usize> for Span {
    fn range(&self) -> Range<usize> {
        (*self).into()
    }
}

impl<'a> Deref for Spanned<&'a str> {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        self.value
    }
}

impl Span {
    pub fn new(start: usize, end: usize) -> Self {
        Self { start, end }
    }

    /// Offset a span by `offset`.
    pub fn offset(&self, offset: usize) -> Self {
        Self {
            start: offset + self.start,
            end: offset + self.end,
        }
    }
}

impl From<Range<usize>> for Span {
    fn from(range: Range<usize>) -> Self {
        Self {
            start: range.start,
            end: range.end,
        }
    }
}

impl From<Span> for Range<usize> {
    fn from(span: Span) -> Self {
        span.start..span.end
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_offset() {
        let span = Span::new(0, 10);
        assert_eq!(span.offset(5), Span::new(5, 15));
    }

    #[test]
    fn test_range() {
        let span = Span::new(0, 10);
        assert_eq!(span.range(), span.into())
    }
}
