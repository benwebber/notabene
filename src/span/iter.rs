use super::Span;

pub struct SpanIterator<'a> {
    text: &'a str,
    cursor: usize,
}

impl<'a> SpanIterator<'a> {
    pub fn new(text: &'a str) -> Self {
        Self { text, cursor: 0 }
    }
}

impl<'a> Iterator for SpanIterator<'a> {
    type Item = Span;

    fn next(&mut self) -> Option<Self::Item> {
        let start = self.text[self.cursor..]
            .char_indices()
            .find(|(_, c)| !c.is_whitespace())
            .map(|(i, _)| self.cursor + i);
        let Some(start) = start else {
            // End of string.
            self.cursor = self.text.len();
            return None;
        };
        let end = self.text[start..]
            .char_indices()
            .find(|(_, c)| c.is_whitespace())
            .map(|(i, _)| start + i)
            .unwrap_or(self.text.len());
        self.cursor = end;
        Some(Span::new(start, end))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_span_iterator() {
        let s = "  foo  foobar  quux  ";
        let spans: Vec<Span> = SpanIterator::new(s).collect();
        assert_eq!(
            spans,
            vec![Span::new(2, 5), Span::new(7, 13), Span::new(15, 19)]
        );
    }
}
