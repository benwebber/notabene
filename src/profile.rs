use std::collections::{HashMap, HashSet};

use crate::rule::Rule;

pub struct Profile {
    pub rules: HashSet<Rule>,
}

impl Profile {
    pub fn new(rules: HashSet<Rule>) -> Self {
        Self { rules }
    }

    pub fn is_enabled(&self, rule: Rule) -> bool {
        self.rules.contains(&rule)
    }
}

impl Default for Profile {
    fn default() -> Self {
        Self::from(Rule::ALL)
    }
}

impl<const N: usize> From<[Rule; N]> for Profile {
    fn from(rules: [Rule; N]) -> Self {
        Self::new(rules.iter().copied().collect())
    }
}
