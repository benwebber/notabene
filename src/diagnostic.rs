//! Linter diagnostics.
use std::path::PathBuf;

use serde::{Deserialize, Serialize};

use crate::rule::Rule;
use crate::span::Index;
use crate::span::Span;
use crate::unist::Position;

/// A rule violation.
#[derive(Debug, PartialEq, Deserialize, Serialize)]
pub struct Diagnostic {
    /// The rule that was violated.
    pub rule: Rule,
    /// Where the violation occurred in the source document.
    pub span: Option<Span>,
    /// The source path, used in reporting.
    pub path: Option<PathBuf>,
}

impl Diagnostic {
    pub fn new(rule: Rule, span: Option<Span>) -> Self {
        Self {
            rule,
            span,
            path: None,
        }
    }

    pub fn code(&self) -> &str {
        self.rule.code()
    }

    pub fn message(&self, source: &str) -> String {
        match self.span {
            Some(span) => {
                let snippet = &source[span.range()];
                self.rule.message().replace("{}", snippet)
            }
            None => self.rule.message().to_string(),
        }
    }

    pub fn position(&self, index: &Index) -> Option<Position> {
        self.span.map(|span| index.position(&span))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use crate::unist::Point;

    #[test]
    fn test_new() {
        assert_eq!(
            Diagnostic::new(Rule::MissingTitle, None),
            Diagnostic {
                rule: Rule::MissingTitle,
                span: None,
                path: None
            }
        );
        assert_eq!(
            Diagnostic::new(Rule::MissingTitle, Some(Span::default())),
            Diagnostic {
                rule: Rule::MissingTitle,
                span: Some(Span::default()),
                path: None
            }
        );
    }

    #[test]
    fn test_code() {
        assert_eq!(
            Diagnostic::new(Rule::MissingTitle, None).code(),
            Rule::MissingTitle.code()
        );
    }

    #[test]
    fn test_message() {
        let source = "";
        let diagnostic = Diagnostic::new(Rule::MissingTitle, None);
        assert_eq!(diagnostic.message(source), Rule::MissingTitle.message());

        let source = "# Changelog";
        let diagnostic = Diagnostic::new(Rule::DuplicateTitle, Some(Span::new(2, 11)));
        assert_eq!(
            diagnostic.message(source),
            Rule::DuplicateTitle.message().replace("{}", "Changelog")
        );
    }

    #[test]
    fn test_position() {
        let index = Index::new("# Changelog");
        let diagnostic = Diagnostic::new(Rule::DuplicateTitle, Some(Span::new(2, 11)));
        assert_eq!(
            diagnostic.position(&index),
            Some(Position::new(Point::new(1, 3, 2), Point::new(1, 12, 11)))
        );
    }
}
