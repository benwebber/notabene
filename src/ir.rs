//! Intermediate representation (IR) of a changelog.
//!
//! Parsing a changelog returns an [`ir::Changelog`](crate::ir::Changelog).
//! This model differs from the [`Changelog`](crate::Changelog) in two significant ways:
//!
//!   * Elements include span information.
//!   * It allows multiple titles and unreleased sections.
//!
//! Preserving span information allows lint checks to include the context around diagnostics.
use crate::span::Span;

use serde::Serialize;

#[derive(Debug, Default, Eq, Hash, PartialEq, Serialize)]
pub(crate) struct Spanned<T> {
    pub span: Span,
    pub value: T,
}

#[derive(Debug, Default, Serialize)]
pub(crate) struct Changelog<'a> {
    pub sections: Vec<Section<'a>>,
    pub broken_links: Vec<Span>,
}

#[derive(Debug, Serialize)]
pub enum Section<'a> {
    Title(Spanned<&'a str>),
    Unreleased(Unreleased<'a>),
    Release(Release<'a>),
    Invalid(InvalidSection),
    // TODO: If Title held something like Vec<Inline>, it would be possible
    // to get rid of InvalidTitle and validate Title in the linter.
    InvalidTitle(InvalidTitle),
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct Unreleased<'a> {
    pub heading_span: Span,
    // TODO: Get span.
    pub url: Option<String>,
    pub changes: Vec<Changes<'a>>,
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct Release<'a> {
    pub heading_span: Span,
    pub version: Spanned<&'a str>,
    // TODO: Get span.
    pub url: Option<String>,
    pub date: Option<Spanned<&'a str>>,
    pub yanked: Option<Spanned<&'a str>>,
    pub changes: Vec<Changes<'a>>,
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct Changes<'a> {
    pub heading_span: Span,
    pub kind: Spanned<&'a str>,
    pub changes: Vec<Spanned<&'a str>>,
}

impl<T> Spanned<T> {
    pub(crate) fn new(span: Span, value: T) -> Self {
        Self { value, span }
    }

    pub(crate) fn into_inner(self) -> T {
        self.value
    }
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct InvalidSection {
    pub heading_span: Span,
}

#[derive(Debug, Default, PartialEq, Serialize)]
pub(crate) struct InvalidTitle {
    pub heading_span: Span,
}
