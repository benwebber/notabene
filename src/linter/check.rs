use crate::diagnostic::Diagnostic;
use crate::ir;
use crate::location::Location;
use crate::rule::Rule;
use crate::span::Span;

pub(crate) trait Check {
    /// Return the `Rule` this check evaluates.
    fn rule(&self) -> Rule;

    /// Return spans that this check matched.
    fn spans(&self) -> &[Span] {
        &[]
    }

    /// Evaluate this check for the entire `Changelog`.
    fn visit_changelog(&mut self, _changelog: &ir::Changelog) {}

    /// Evaluate this check for a `Section`.
    fn visit_section(&mut self, _section: &ir::Section) {}

    /// Evaluate this check for a `Changes`.
    fn visit_changes(&mut self, _changes: &ir::Changes) {}

    /// Publish diagnostics generated by this check.
    fn diagnostics(&self) -> Vec<Diagnostic> {
        self.spans()
            .iter()
            .map(|span| Diagnostic::new(self.rule(), Some(Location::Span(*span))))
            .collect()
    }
}
