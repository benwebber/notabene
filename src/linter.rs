//! Linter implementation.
use std::collections::HashSet;

use time::Date;
use time::macros::format_description;

use crate::diagnostic::Diagnostic;
use crate::ir::{Changelog, Changes, Section};
use crate::profile::Profile;
use crate::rule::Rule;
use crate::span::Span;

mod check;

use check::Check;

pub fn lint(changelog: &Changelog, profile: &Profile) -> Vec<Diagnostic> {
    let mut diagnostics = Vec::new();

    let mut checks: Vec<_> = checks()
        .into_iter()
        .filter(|check| profile.is_enabled(check.rule()))
        .collect();

    for section in &changelog.sections {
        for check in checks.iter_mut() {
            check.visit_section(section);
            match section {
                Section::Unreleased(unreleased) => {
                    for changes in &unreleased.changes {
                        check.visit_changes(changes);
                    }
                }
                Section::Release(release) => {
                    for changes in &release.changes {
                        check.visit_changes(changes);
                    }
                }
                _ => {}
            }
        }
    }

    for check in checks.iter_mut() {
        diagnostics.append(&mut check.diagnostics());
    }

    diagnostics
}

fn checks() -> Vec<Box<dyn Check>> {
    vec![
        // E000 Document
        Box::new(MissingTitleCheck::default()),
        Box::new(DuplicateTitleCheck::default()),
        Box::new(UnreleasedNotFirstCheck::default()),
        //// E100 Unreleased
        Box::new(MissingUnreleasedCheck::default()),
        Box::new(DuplicateUnreleasedCheck::default()),
        //// E200 Release
        Box::new(InvalidDateCheck::default()),
        Box::new(InvalidYankedCheck::default()),
        //// E300 Changes
        Box::new(InvalidChangeTypeCheck::default()),
        Box::new(DuplicateChangeTypeCheck::default()),
        //// E400 Content
        Box::new(EmptySectionCheck::default()),
    ]
}

// E000 Document
// =============

#[derive(Default)]
struct MissingTitleCheck {
    found: bool,
}

impl Check for MissingTitleCheck {
    fn rule(&self) -> Rule {
        Rule::MissingTitle
    }

    fn visit_section(&mut self, section: &Section) {
        if self.found {
            return;
        }
        if matches!(section, Section::Title(_)) {
            self.found = true;
        }
    }

    fn diagnostics(&self) -> Vec<Diagnostic> {
        if self.found {
            vec![]
        } else {
            vec![Diagnostic::new(self.rule(), None)]
        }
    }
}

#[derive(Default)]
struct DuplicateTitleCheck {
    spans: Vec<Span>,
    found: bool,
}

impl Check for DuplicateTitleCheck {
    fn rule(&self) -> Rule {
        Rule::DuplicateTitle
    }

    fn spans(&self) -> &[Span] {
        self.spans.as_slice()
    }

    fn visit_section(&mut self, section: &Section) {
        if let Section::Title(spanned) = section {
            if self.found {
                self.spans.push(spanned.span);
            } else {
                self.found = true;
            }
        }
    }
}

// E100 Unreleased
// ===============

#[derive(Default)]
struct MissingUnreleasedCheck {
    found: bool,
}

impl Check for MissingUnreleasedCheck {
    fn rule(&self) -> Rule {
        Rule::MissingUnreleased
    }

    fn visit_section(&mut self, section: &Section) {
        if self.found {
            return;
        }
        if matches!(section, Section::Unreleased(_)) {
            self.found = true;
        }
    }

    fn diagnostics(&self) -> Vec<Diagnostic> {
        if self.found {
            vec![]
        } else {
            vec![Diagnostic::new(self.rule(), None)]
        }
    }
}

#[derive(Default)]
struct DuplicateUnreleasedCheck {
    spans: Vec<Span>,
    found: bool,
}

impl Check for DuplicateUnreleasedCheck {
    fn rule(&self) -> Rule {
        Rule::DuplicateUnreleased
    }

    fn spans(&self) -> &[Span] {
        self.spans.as_slice()
    }

    fn visit_section(&mut self, section: &Section) {
        if let Section::Unreleased(unreleased) = section {
            if self.found {
                self.spans.push(unreleased.heading_span);
            } else {
                self.found = true;
            }
        }
    }
}

#[derive(Default)]
struct UnreleasedNotFirstCheck {
    spans: Vec<Span>,
    found: bool,
}

impl Check for UnreleasedNotFirstCheck {
    fn rule(&self) -> Rule {
        Rule::UnreleasedNotFirst
    }

    fn spans(&self) -> &[Span] {
        self.spans.as_slice()
    }

    fn visit_section(&mut self, section: &Section) {
        if self.found {
            return;
        }
        match section {
            Section::Unreleased(_) => self.found = true,
            Section::Release(release) => self.spans.push(release.heading_span),
            _ => {}
        }
    }
}

// E200 Release
// ============

#[derive(Default)]
struct InvalidDateCheck {
    spans: Vec<Span>,
}

impl Check for InvalidDateCheck {
    fn rule(&self) -> Rule {
        Rule::InvalidDate
    }

    fn spans(&self) -> &[Span] {
        self.spans.as_slice()
    }

    fn visit_section(&mut self, section: &Section) {
        let format = format_description!("[year]-[month]-[day]");
        if let Section::Release(release) = section {
            if let Some(spanned) = &release.date {
                if Date::parse(&spanned.value, &format).is_err() {
                    self.spans.push(spanned.span);
                }
            }
        }
    }
}

#[derive(Default)]
struct InvalidYankedCheck {
    spans: Vec<Span>,
}

impl Check for InvalidYankedCheck {
    fn rule(&self) -> Rule {
        Rule::InvalidYanked
    }

    fn spans(&self) -> &[Span] {
        self.spans.as_slice()
    }

    fn visit_section(&mut self, section: &Section) {
        if let Section::Release(release) = section {
            if let Some(spanned) = &release.yanked {
                if spanned.value != "[YANKED]" {
                    self.spans.push(spanned.span);
                }
            }
        }
    }
}

// E300 Changes
// ============

#[derive(Default)]
struct InvalidChangeTypeCheck {
    spans: Vec<Span>,
}

impl Check for InvalidChangeTypeCheck {
    fn rule(&self) -> Rule {
        Rule::InvalidChangeType
    }

    fn spans(&self) -> &[Span] {
        self.spans.as_slice()
    }

    fn visit_changes(&mut self, changes: &Changes) {
        if !matches!(
            changes.kind.value.as_str(),
            "Added" | "Changed" | "Deprecated" | "Fixed" | "Removed" | "Security"
        ) {
            self.spans.push(changes.kind.span);
        }
    }
}

#[derive(Default)]
struct DuplicateChangeTypeCheck {
    spans: Vec<Span>,
    seen: HashSet<String>,
}

impl Check for DuplicateChangeTypeCheck {
    fn rule(&self) -> Rule {
        Rule::DuplicateChangeType
    }

    fn spans(&self) -> &[Span] {
        self.spans.as_slice()
    }

    fn visit_section(&mut self, _section: &Section) {
        self.seen.clear();
    }

    fn visit_changes(&mut self, changes: &Changes) {
        if !self.seen.insert(changes.kind.value.clone()) {
            self.spans.push(changes.kind.span);
        }
    }
}

// E400 Content
// ============

#[derive(Default)]
struct EmptySectionCheck {
    spans: Vec<Span>,
}

impl Check for EmptySectionCheck {
    fn rule(&self) -> Rule {
        Rule::EmptySection
    }

    fn spans(&self) -> &[Span] {
        self.spans.as_slice()
    }

    fn visit_section(&mut self, section: &Section) {
        match section {
            Section::Unreleased(unreleased) => {
                if unreleased.changes.is_empty() {
                    self.spans.push(unreleased.heading_span);
                }
            }
            Section::Release(release) => {
                if release.changes.is_empty() {
                    self.spans.push(release.heading_span);
                }
            }
            _ => {}
        }
    }

    fn visit_changes(&mut self, changes: &Changes) {
        if changes.changes.is_empty() {
            self.spans.push(changes.heading_span)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_yaml_snapshot;

    use crate::ir::*;
    use crate::profile::Profile;
    use crate::span::Span;

    #[test]
    fn test_missing_title() {
        let profile = Profile::new(&[Rule::MissingTitle]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![Section::Title(Spanned::<String>::default())],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }

    #[test]
    fn test_duplicate_title() {
        let profile = Profile::new(&[Rule::DuplicateTitle]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![
                Section::Title(Spanned::<String>::default()),
                Section::Title(Spanned::new(Span::new(2, 11), "Changelog".to_string())),
            ],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }

    #[test]
    fn test_missing_unreleased() {
        let profile = Profile::new(&[Rule::MissingUnreleased]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![Section::Unreleased(Unreleased::default())],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }

    #[test]
    fn test_unreleased_not_first() {
        let profile = Profile::new(&[Rule::UnreleasedNotFirst]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![
                Section::Release(Release {
                    heading_span: Span::new(1, usize::MAX),
                    ..Default::default()
                }),
                Section::Unreleased(Unreleased::default()),
            ],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }

    #[test]
    fn test_duplicate_unreleased() {
        let profile = Profile::new(&[Rule::DuplicateUnreleased]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![
                Section::Unreleased(Unreleased::default()),
                Section::Unreleased(Unreleased {
                    heading_span: Span::new(0, 17),
                    ..Default::default()
                }),
            ],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }

    #[test]
    fn test_invalid_date() {
        let profile = Profile::new(&[Rule::InvalidDate]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![
                Section::Release(Release {
                    ..Default::default()
                }),
                Section::Release(Release {
                    date: Some(Spanned::new(Span::new(0, 9), "2038-01-19".to_string())),
                    ..Default::default()
                }),
                Section::Release(Release {
                    date: Some(Spanned::new(Span::new(1, 10), "2001-01-00".to_string())),
                    ..Default::default()
                }),
                Section::Release(Release {
                    date: Some(Spanned::new(Span::new(2, 5), "foo".to_string())),
                    ..Default::default()
                }),
            ],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }

    #[test]
    fn test_invalid_yanked() {
        let profile = Profile::new(&[Rule::InvalidYanked]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![
                Section::Release(Release {
                    ..Default::default()
                }),
                Section::Release(Release {
                    yanked: Some(Spanned::new(Span::new(0, 9), "[YANKED]".to_string())),
                    ..Default::default()
                }),
                Section::Release(Release {
                    yanked: Some(Spanned::new(Span::new(1, 10), "[ZANKED]".to_string())),
                    ..Default::default()
                }),
            ],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }

    #[test]
    fn test_invalid_change_type() {
        let profile = Profile::new(&[Rule::InvalidChangeType]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![
                Section::Unreleased(Unreleased {
                    changes: vec![
                        Changes {
                            kind: Spanned::new(Span::new(0, 0), "Added".to_string()),
                            ..Default::default()
                        },
                        Changes {
                            kind: Spanned::new(Span::new(1, usize::MAX), "Foo".to_string()),
                            ..Default::default()
                        },
                    ],
                    ..Default::default()
                }),
                Section::Release(Release {
                    changes: vec![
                        Changes {
                            kind: Spanned::new(Span::new(0, 0), "Added".to_string()),
                            ..Default::default()
                        },
                        Changes {
                            kind: Spanned::new(Span::new(2, usize::MAX), "Foo".to_string()),
                            ..Default::default()
                        },
                    ],
                    ..Default::default()
                }),
            ],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }

    #[test]
    fn test_duplicate_change_type() {
        let profile = Profile::new(&[Rule::DuplicateChangeType]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![
                Section::Unreleased(Unreleased {
                    changes: vec![
                        Changes {
                            kind: Spanned::new(Span::new(0, 0), "Added".to_string()),
                            ..Default::default()
                        },
                        Changes {
                            kind: Spanned::new(Span::new(1, usize::MAX), "Added".to_string()),
                            ..Default::default()
                        },
                    ],
                    ..Default::default()
                }),
                Section::Release(Release {
                    changes: vec![
                        Changes {
                            kind: Spanned::new(Span::new(0, 0), "Added".to_string()),
                            ..Default::default()
                        },
                        Changes {
                            kind: Spanned::new(Span::new(2, usize::MAX), "Added".to_string()),
                            ..Default::default()
                        },
                    ],
                    ..Default::default()
                }),
            ],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }

    #[test]
    fn test_empty_section() {
        let profile = Profile::new(&[Rule::EmptySection]);

        let changelog = Changelog::default();
        assert_yaml_snapshot!(lint(&changelog, &profile));

        let changelog = Changelog {
            sections: vec![
                // No changes.
                Section::Unreleased(Unreleased {
                    heading_span: Span::new(1, usize::MAX),
                    ..Default::default()
                }),
                Section::Unreleased(Unreleased {
                    changes: vec![
                        Changes {
                            kind: Spanned::new(Span::new(0, 0), "Added".to_string()),
                            changes: vec![Spanned::new(Span::new(0, 0), "Add foo".to_string())],
                            ..Default::default()
                        },
                        // Empty changes.
                        Changes {
                            heading_span: Span::new(2, usize::MAX),
                            ..Default::default()
                        },
                    ],
                    ..Default::default()
                }),
                // No changes.
                Section::Release(Release {
                    heading_span: Span::new(3, usize::MAX),
                    ..Default::default()
                }),
                Section::Release(Release {
                    changes: vec![
                        Changes {
                            kind: Spanned::new(Span::new(0, 0), "Added".to_string()),
                            changes: vec![Spanned::new(Span::new(0, 0), "Add foo".to_string())],
                            ..Default::default()
                        },
                        // Empty changes.
                        Changes {
                            heading_span: Span::new(4, usize::MAX),
                            ..Default::default()
                        },
                    ],
                    ..Default::default()
                }),
            ],
            ..Default::default()
        };
        assert_yaml_snapshot!(lint(&changelog, &profile));
    }
}
