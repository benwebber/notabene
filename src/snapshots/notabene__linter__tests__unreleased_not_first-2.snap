---
source: src/linter.rs
expression: "lint(&changelog, &profile)"
---
- rule: UnreleasedNotFirst
  span:
    start: 1
    end: 18446744073709551615
  path: ~
